### PR TITLE
 seed: stop span before returning job outcome 

### PR DESCRIPTION
**Pull request by bboozzoo**
Make sure to stop the time span before returning job's outcome.

This was identified as a data race where the DeviceManager.ensureSeeded() would call seed20.LoadMeta(), which internally loads seed meta and launches load-meta jobs. When exuction goes back to in ensureSeeded(), the timings would be saved to state, which causes flattening of all time spans. Since load-meta jobs were running, and the code in seed20.doLoadMeta() only synchronized with results from outcomeCh, there was no guarantee that TimeSpan.Stop() ran before we exited the function.

Backtrace from go test -race:

```
WARNING: DATA RACE                                                                                                                                                                                                                                                                                                                                                           16:11:48 [54/9269]
Read at 0x00c0000f06c8 by goroutine 14:
  runtime.racereadrange()
      <autogenerated>:1 +0x1b
  github.com/snapcore/snapd/timings.flattenRecursive()
      /home/maciek/work/canonical/snapd/timings/state.go:113 +0x53c
  github.com/snapcore/snapd/timings.flattenRecursive()
      /home/maciek/work/canonical/snapd/timings/state.go:113 +0x53c
  github.com/snapcore/snapd/timings.(*Timings).flatten()
      /home/maciek/work/canonical/snapd/timings/state.go:87 +0x22c
  github.com/snapcore/snapd/timings.(*Timings).Save()
      /home/maciek/work/canonical/snapd/timings/state.go:140 +0x152
  github.com/snapcore/snapd/overlord/devicestate.(*DeviceManager).ensureSeeded()
      /home/maciek/work/canonical/snapd/overlord/devicestate/devicemgr.go:1007 +0x6d7
  github.com/snapcore/snapd/overlord/devicestate.(*DeviceManager).Ensure()
      /home/maciek/work/canonical/snapd/overlord/devicestate/devicemgr.go:1727 +0x4a
  github.com/snapcore/snapd/overlord.(*StateEngine).Ensure()
      /home/maciek/work/canonical/snapd/overlord/stateengine.go:147 +0x333
  github.com/snapcore/snapd/overlord.(*Overlord).settle()
      /home/maciek/work/canonical/snapd/overlord/overlord.go:553 +0x285
  github.com/snapcore/snapd/overlord.(*Overlord).Settle()
      /home/maciek/work/canonical/snapd/overlord/overlord.go:596 +0x15f1
  github.com/snapcore/snapd/overlord/devicestate_test.(*firstBoot20Suite).testPopulateFromSeedCore20ValidationSetTracking()
      /home/maciek/work/canonical/snapd/overlord/devicestate/firstboot20_test.go:1259 +0x15af
  github.com/snapcore/snapd/overlord/devicestate_test.(*firstBoot20Suite).TestPopulateFromSeedCore20ValidationSetTrackingFailsUnmetCriterias()
      /home/maciek/work/canonical/snapd/overlord/devicestate/firstboot20_test.go:1386 +0xa11
  runtime.call16()
      /snap/go/10029/src/runtime/asm_amd64.s:701 +0x48
  reflect.Value.Call()
      /snap/go/10029/src/reflect/value.go:339 +0xd7
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /home/maciek/work/canonical/snapd/vendor/gopkg.in/check.v1/check.go:775 +0xa71
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /home/maciek/work/canonical/snapd/vendor/gopkg.in/check.v1/check.go:669 +0xe5

Previous write at 0x00c0000f06c8 by goroutine 74:
  github.com/snapcore/snapd/timings.(*Span).Stop()
      /home/maciek/work/canonical/snapd/timings/timings.go:119 +0xee
  github.com/snapcore/snapd/seed.(*seed20).doLoadMeta.func5.1()
      /home/maciek/work/canonical/snapd/seed/seed20.go:586 +0x39
  runtime.deferreturn()
      /snap/go/10029/src/runtime/panic.go:436 +0x32

Goroutine 14 (running) created at:
  gopkg.in/check%2ev1.(*suiteRunner).forkCall()
      /home/maciek/work/canonical/snapd/vendor/gopkg.in/check.v1/check.go:666 +0x584
  gopkg.in/check%2ev1.(*suiteRunner).forkTest()
      /home/maciek/work/canonical/snapd/vendor/gopkg.in/check.v1/check.go:757 +0x164
  gopkg.in/check%2ev1.(*suiteRunner).runTest()
      /home/maciek/work/canonical/snapd/vendor/gopkg.in/check.v1/check.go:812 +0x414
  gopkg.in/check%2ev1.(*suiteRunner).run()
      /home/maciek/work/canonical/snapd/vendor/gopkg.in/check.v1/check.go:618 +0x3bf
  gopkg.in/check%2ev1.Run()
      /home/maciek/work/canonical/snapd/vendor/gopkg.in/check.v1/run.go:92 +0x49
  gopkg.in/check%2ev1.RunAll()
      /home/maciek/work/canonical/snapd/vendor/gopkg.in/check.v1/run.go:84 +0x127
  gopkg.in/check%2ev1.TestingT()
      /home/maciek/work/canonical/snapd/vendor/gopkg.in/check.v1/run.go:72 +0x571
  github.com/snapcore/snapd/overlord/devicestate_test.TestDeviceManager()
      /home/maciek/work/canonical/snapd/overlord/devicestate/devicestate_test.go:80 +0x2e
  testing.tRunner()
      /snap/go/10029/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /snap/go/10029/src/testing/testing.go:1486 +0x47

Goroutine 74 (finished) created at:
  github.com/snapcore/snapd/seed.(*seed20).doLoadMeta()
      /home/maciek/work/canonical/snapd/seed/seed20.go:585 +0x664
  github.com/snapcore/snapd/seed.(*seed20).LoadMeta()
      /home/maciek/work/canonical/snapd/seed/seed20.go:703 +0x290
  github.com/snapcore/snapd/overlord/devicestate.(*DeviceManager).populateStateFromSeedImpl.func2()
      /home/maciek/work/canonical/snapd/overlord/devicestate/firstboot.go:170 +0x8c
  github.com/snapcore/snapd/timings.Run()
      /home/maciek/work/canonical/snapd/timings/helpers.go:26 +0x84
  github.com/snapcore/snapd/overlord/devicestate.(*DeviceManager).populateStateFromSeedImpl()
      /home/maciek/work/canonical/snapd/overlord/devicestate/firstboot.go:169 +0x40c
  github.com/snapcore/snapd/overlord/devicestate.(*DeviceManager).populateStateFromSeedImpl-fm()
      <autogenerated>:1 +0x59
  github.com/snapcore/snapd/overlord/devicestate.(*DeviceManager).ensureSeeded.func1()
      /home/maciek/work/canonical/snapd/overlord/devicestate/devicemgr.go:991 +0x74
```

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

### CODE REVIEW

File: seed/seed20.go [hunk 1]
```diff
@@ -580,9 +580,15 @@ func (s *seed20) doLoadMeta(handler SnapHandler, tm timings.Measurer) error {
...
+	var wg sync.WaitGroup
+	wg.Add(njobs)
 	for j := 1; j <= njobs; j++ {
 		jtm := tm.StartSpan(fmt.Sprintf("do-load-meta[%d]", j), fmt.Sprintf("snap metadata loading job #%d", j))
 		go func() {
+			// defers are LIFO, make sure that time snap is stopped
+			// before we let the parent know that the goroutine is
+			// done
+			defer wg.Done()
 			defer jtm.Stop()
```
Comment (pedronis) [inline seed/seed20.go:592]: shouldn't we instead move the Stop before we send things on the channels? it's a bit of duplication, but adding a wait group just for dealing with timings seems also a bit overkill?

Comment (olivercalder) [reply seed/seed20.go:592]: That would seem to solve the problem as well. The wait group is more explicit as to why it is safe, but moving the `jtm.Stop()` call is lighter-weight.

Comment (olivercalder) [inline seed/seed20.go:592]: How about this:

```go
var retErr error
defer func() {
    outcomesCh <- retErr
}()
defer jtm.Stop()
...
        if err == errSkipped {
            continue
        }
        retErr = err
        return
...
```

This way, we use defer to ensure `jtm.Stop()` is always closed, no matter which return is taken, and the defer stack ensures `jtm.Stop()` is called before sending over the channel.

**Review by olivercalder — COMMENTED** [review]
Thanks! I propose a third approach, which doesn't require a WaitGroup but preserves the safety of not needing to call `jtm.Stop()` manually from multiple places. I agree that the WaitGroup method is the most explicit as to why safety is guaranteed, though.

File: seed/seed20.go (hunk 1 above)
Comment (bboozzoo) [reply seed/seed20.go:591]: and pushed

**Review by pedronis — APPROVED** [review]
thank you

File: seed/seed20.go [hunk 2]
```diff
@@ -583,6 +583,11 @@ func (s *seed20) doLoadMeta(handler SnapHandler, tm timings.Measurer) error {
 	for j := 1; j <= njobs; j++ {
 		jtm := tm.StartSpan(fmt.Sprintf("do-load-meta[%d]", j), fmt.Sprintf("snap metadata loading job #%d", j))
 		go func() {
+			var jobErr error
+			// defers are LIFO, make sure that time snap is stopped
+			// before we let the parent know that the goroutine is
+			// done
+			defer func() { outcomesCh <- jobErr }()
```
Comment (olivercalder) [inline seed/seed20.go:590]: Nice comment, thanks!

**Review by olivercalder — APPROVED** [review]
Thanks for the changes, looks great!

**Pull request merged**